### PR TITLE
Type dictionary

### PR DESCRIPTION
**Pull request by mdornaus**

### CODE REVIEW

File: DashboardClient/DashboardClient.cpp [hunk 1]
```diff
@@ -383,14 +422,18 @@ namespace Umati
 			std::map<std::shared_ptr<const ModelOpcUa::Node>, nlohmann::json> &valueMap,
 			std::mutex &valueMap_mutex)
 		{
-			// LOG(INFO) << "subscribeValues "   << pNode->NodeId.Uri << ";" << pNode->NodeId.Id;
+			LOG(INFO) << "subscribeValues "   << pNode->NodeId.Uri << ";" << pNode->NodeId.Id;
+			for(auto value : m_subscribedValues){
```
Comment (Kantiran91) [inline DashboardClient/DashboardClient.cpp:426]: Can you make a comment, why the loop is nessasary? I think you check if the values is already subscribed? Is this correct?

File: OpcUaClient/Converter/UaDataValueToJsonValue.cpp [hunk 2]
```diff
@@ -1,490 +1,456 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
+ /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- *
- * Copyright 2019-2022 (c) Christian von Arnim, ISW University of Stuttgart (for umati and VDW e.V.)
+ * 
+ * Copyright 2019-2021 (c) Christian von Arnim, ISW University of Stuttgart (for umati and VDW e.V.)
```
Comment (Kantiran91) [inline OpcUaClient/Converter/UaDataValueToJsonValue.cpp:5]: can you change this to 22 back?

File: OpcUaClient/OpcUaClient.cpp [hunk 3]
```diff
@@ -371,15 +667,15 @@ void OpcUaClient::fillNamespaceCache(const std::vector<std::string> &uaNamespace
 }
 
 void OpcUaClient::updateCustomDataTypesNamespace(std::string namespaceURI, std::size_t namespaceIndex) {
-  if (namespaceURI == "http://opcfoundation.org/UA/Machinery/Result/") {
+  /*if (namespaceURI == "http://opcfoundation.org/UA/Machinery/Result/") {
```
Comment (Kantiran91) [inline OpcUaClient/OpcUaClient.cpp:670]: please remove old code

**Review by Kantiran91 — CHANGES_REQUESTED** [review]
@mdornaus I found some smaller issues, where I get in trubble in the understanding.

File: DashboardClient/Converter/ModelToJson.cpp [hunk 4]
```diff
@@ -105,14 +105,17 @@ namespace Umati {
 			//Set ofBaseDataVariableType somewhere?
 			bool ModelToJson::isBaseDataVariableType(
 					const std::shared_ptr<const ModelOpcUa::SimpleNode> &pSimpleNode) {
-				return (pSimpleNode->SpecifiedTypeNodeId.Uri == "" 
+				// FIX_BEGIN (FIX_5_PART_2)
```
[the remark this answers is not in the document]
Comment (wlkrm) [reply DashboardClient/Converter/ModelToJson.cpp:108]: Also remove all manually implemented CustomTypes and so on. Everything like https://github.com/umati/Dashboard-OPCUA-Client/blob/development/OpcUaClient/Converter/CustomDataTypes/types_machinery_result_generated.c can be removed.

File: DashboardClient/DashboardClient.cpp [hunk 5]
```diff
@@ -145,7 +145,29 @@ namespace Umati
...
+					//
+					//
+					auto pSimpleNode = std::dynamic_pointer_cast<const ModelOpcUa::SimpleNode>(pNode);
+					if(pSimpleNode) {
+						//std::cout << pSimpleNode->NodeId << "\n";
+						auto values = pDataSetStorage->values;
+						for(auto it1 : values) {
+							auto pSimpleNode1 = std::dynamic_pointer_cast<const ModelOpcUa::SimpleNode>(it1.first);
+								if(pSimpleNode1->NodeId == pSimpleNode->NodeId) {
+									// DEBUG_BEGIN in case we want to see the different pointer addresses.
```
Comment (wlkrm) [inline DashboardClient/DashboardClient.cpp:160]: Why not use LOG(DEBUG) ?

File: OpcUaClient/OpcUaClient.hpp [hunk 6]
```diff
@@ -112,6 +120,47 @@ class OpcUaClient : public Dashboard::IDashboardDataClient {
...
+				{ "ua:ExtensionObject", UA_TYPES_EXTENSIONOBJECT },
+				{ "ua:DataValue", UA_TYPES_DATAVALUE },
+				{ "ua:Variant", UA_TYPES_VARIANT },
+				{ "ua:DiagnosticInfo", UA_TYPES_DIAGNOSTICINFO },
+				{ "ua:SimpleAttributeOperand", UA_TYPES_SIMPLEATTRIBUTEOPERAND},
+				{ "ua:ContentFilter", UA_TYPES_CONTENTFILTER},
+				{ "ua:Range", UA_TYPES_RANGE},
+				{ "ua:EUInformation", UA_TYPES_EUINFORMATION}
+
+				/* TODO ???
```
Comment (wlkrm) [inline OpcUaClient/OpcUaClient.hpp:155]: What about these other Types?

**Review by wlkrm — CHANGES_REQUESTED** [review]
Should be fine, as I think it's mainly the code I wrote last year.
SOmethings are still left to do:
* Remove everything that is CustomTypes specific (such as https://github.com/umati/Dashboard-OPCUA-Client/tree/development/OpcUaClient/Converter/CustomDataTypes)
* Especially this file should be significantly reduces in line numbers and complexity (https://github.com/umati/Dashboard-OPCUA-Client/blob/development/OpcUaClient/Converter/UaDataValueToJsonValue.cpp)
* You could have just rebased my commits to keep track of the developer (me) to blame if something does not work. But it doesn't really matter in this case. 
* Wait for go of @GoetzGoerisch if we really want to break with datahub compatibility today

**Comment by GoetzGoerisch** [comment]
> * Wait for go of @GoetzGoerisch if we really want to break with datahub compatibility today

Ok, from my side.

**Review by GoetzGoerisch — CHANGES_REQUESTED** [review]
Running this version breaks the correct publishing for e.g. MachineTools

**Comment by wlkrm** [comment]
LGTM, after the remaining comments are resolved.

File: DashboardClient/Converter/ModelToJson.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mdornaus) [reply DashboardClient/Converter/ModelToJson.cpp]: Rmoved old code and generated classes.

File: DashboardClient/DashboardClient.cpp (hunk 1 above)
Comment (mdornaus) [reply DashboardClient/DashboardClient.cpp]: Yes.

File: OpcUaClient/OpcUaClient.cpp (hunk 3 above)
Comment (mdornaus) [reply OpcUaClient/OpcUaClient.cpp]: Removed

File: OpcUaClient/Converter/UaDataValueToJsonValue.cpp (hunk 2 above)
Comment (mdornaus) [reply OpcUaClient/Converter/UaDataValueToJsonValue.cpp]: Changed Back.

File: DashboardClient/DashboardClient.cpp (hunk 5 above)
Comment (mdornaus) [reply DashboardClient/DashboardClient.cpp:157]: Removed and LOG(DEBUG used). Also removed old classes.

**Review by Kantiran91 — APPROVED** [review]

File: OpcUaClient/OpcUaClient.hpp (hunk 6 above)
Comment (mdornaus) [reply OpcUaClient/OpcUaClient.hpp:155]: The aggregating Types are handled different. I suggest to leave it this way until we run into problems.

**Review by GoetzGoerisch — APPROVED** [review]
LGTM, deployed for sample server currently

**Review by wlkrm — APPROVED** [review]

**Pull request merged**